### PR TITLE
LPS-186307 Periodically clean up files in 'new' store area

### PR DESCRIPTION
**Pull request by adolfopa**
# What is this about?

https://issues.liferay.com/browse/LPS-186307

Here is the rest of the code to manage new files using multiple store areas. This PR contains the background task to periodically clean up new (but orphan) files from the store and to evict them to the 'live' area when they're old enough.

### CODE REVIEW

**Review by marcogalluzzi — CHANGES_REQUESTED** [review]
I would change two general things.

First, IMO using the word *eviction* is not good anymore. When starting with these changes *eviction* meant moving a file from LIVE area to the DELETE area. If we have a look to method `_processStoreArea`, the word *eviction* is used for processing a file, this is, deleting from the DELETED area or copying to LIVE + deleting from NEW area.

I would change it to something more general like “processing” so we could have in the configuration these descriptions:

```
store-area-minimum-age=Store Area Minimum Age
store-area-minimum-age-help=Set the minimum age of files (in days) before they can be cleaned up.
store-area-processing-quota=Store Area Processing Quota
store-area-processing-quota-help=Set how many files to process during a cleanup.
```

Second, having a look to the method `_processStoreArea`, I would not make actual work in a `predicate` parameter because it is harder to understand what the method is doing. I would change this parameter to be the `task`:

* In one case the `task` would be deleting from DELETE area
* In the other case the `task` would be copying to LIVE + deleting from NEW area

File: modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/CleanUpStoreAreasSchedulerJobConfiguration.java [hunk 1]
```diff
@@ -74,18 +75,54 @@ protected void activate(Map<String, Object> properties) {
...
+		if (pos == -1) {
+			return true;
+		}
+
+		int fileVersionsCount = _dlFileVersionLocalService.getFileVersionsCount(
+			companyId, name.substring(pos + 1));
+
+		if (fileVersionsCount > 0) {
+			return true;
+		}
```
[the remark this answers is not in the document]
Comment (adolfopa) [reply modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/CleanUpStoreAreasSchedulerJobConfiguration.java:115]: No, SF complained about it.

File: modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/CleanUpStoreAreasSchedulerJobConfiguration.java [hunk 2]
```diff
@@ -57,7 +58,7 @@
 	@Override
 	public UnsafeRunnable<Exception> getJobExecutorUnsafeRunnable() {
 		return () -> _companyLocalService.forEachCompanyId(
-			this::_cleanUpDeletedStoreArea);
+			this::_cleanUpStorageAreas);
 	}
```
[the remark this answers is not in the document]
Comment (adolfopa) [reply modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/CleanUpStoreAreasSchedulerJobConfiguration.java:62]: You're right.

**Comment by adolfopa** [comment]
@marcogalluzzi I disagree with both of your suggestions :-)
1. "Evicted" is roughly equivalent to "expelled", which reflects what we're doing (deletion is a kind of permanent expelling).
2. About passing a "task" identifier, I think it actually makes the code more convoluted; instead of a per task predicate, you'd have a hairy conditional that depends on the storage area, the predicate and the result of the copy operations. I'm not saying the current approach is great, but at least we simplify the conditions a little bit by containing them in the specific method for each area.

**Comment by adolfopa** [comment]
@marcogalluzzi I've squashed the changes into https://github.com/liferay-lima/liferay-portal/pull/4358/commits/ee18401349948689f88e858ab2039a86f3625bf2

**Comment by marcogalluzzi** [comment]
> @marcogalluzzi I disagree with both of your suggestions :-)
> 
> 1. "Evicted" is roughly equivalent to "expelled", which reflects what we're doing (deletion is a kind of permanent expelling).
> 2. About passing a "task" identifier, I think it actually makes the code more convoluted; instead of a per task predicate, you'd have a hairy conditional that depends on the storage area, the predicate and the result of the copy operations. I'm not saying the current approach is great, but at least we simplify the conditions a little bit by containing them in the specific method for each area.

About 1)
With my proposition I just wanted to simplify the understanding of the Store Area and its configuration. 
For instance, reading the method `_processStoreArea`, the use of the term "eviction" is a bit confusing. Do we refer to "evict" to what are we doing to files in the DELETED/NEW area (removing from there) or to what we have done to deleted files (moved to DELETED area)?
If I read variables `evictedBlobQuota` and `evictedBlobsCount` it seems to refer to the first case. If I read variable `evictionInstant` it seems to refer to the second case, which cannot be applied to files in NEW area (they haven't been expelled from anywhere).

Anyway, at least this description should be changed IMO because it doesn't apply to the NEW store area:

```
store-area-eviction-age-help=Set the age of deleted files (in days) before they can be permanently deleted.
```

About 2) I don't understand how the proposal make the code more convoluted.

```java
	private String _processStoreArea(
		StoreArea storeArea, long companyId, int evictionQuota,
		BiConsumer<Blob, StorageBatch> task, TemporalAmount temporalAmount,
		String startOffset) {
	...
				if (evictionInstant.isBefore(instant)) {
					task.accept(blob, storageBatch);

					evictedBlobQuota--;
					evictedBlobsCount++;
				}
	...
	}
```

```java
	@Override
	public String cleanUpDeletedStoreArea(
		long companyId, int deletionQuota, Predicate<String> predicate,
		TemporalAmount temporalAmount, String startOffset) {

		return _processStoreArea(
			StoreArea.DELETED, companyId, deletionQuota,
			(blob, storageBatch) -> {
				if (predicate.test(blob.getName())) {
					storageBatch.delete(blob.getBlobId());
				}},
			temporalAmount, startOffset);
	}

	@Override
	public String cleanUpNewStoreArea(
		long companyId, int evictionQuota, Predicate<String> predicate,
		TemporalAmount temporalAmount, String startOffset) {

		return _processStoreArea(
			StoreArea.NEW, companyId, evictionQuota,
			(blob, storageBatch) -> {
				StoreArea currentStoreArea = StoreArea.LIVE;

				if (predicate.test(blob.getName())) {
					currentStoreArea = StoreArea.DELETED;
				}

				if (copy(blob.getName(), StoreArea.NEW.relocate(blob.getName(), currentStoreArea))) {
					storageBatch.delete(blob.getBlobId());
				};
			},
			temporalAmount, startOffset);
	}
```

**Comment by adolfopa** [comment]
@marcogalluzzi About 1), the `evictionInstant` refers to the point in time when the blob will be evicted, not the time it was deleted from DM. The `evictionTime` instant is comparted with the current instant, and if it is time to evict the element (my moving it to other area or deleting it) then it will do it.

About 2) I though you were suggesting something much more crude: passing a task identifier and resolving the logic internally with a bunch of conditionals. In any case, it would need to be a BiFunction, as the blob counters need to be updated only if the task was successful (i.e. the copy was succesful). And then we're back to square 1, except that you now have moved that one line of code (batch.delete) into those functions. I still don't see the advantage over keeping the deletion action in the main method.

**Comment by marcogalluzzi** [comment]
@adolfopa I agree. The only pending thing is... do you think this description is OK?

```
store-area-eviction-age-help=Set the age of deleted files (in days) before they can be permanently deleted.
```

Would it be more correctly something like:

```
store-area-eviction-age-help=Set the age of files (in days) before they can be evicted.
```

**Comment by adolfopa** [comment]
@marcogalluzzi Sorry, I didn't notice your suggestion about the lang key. You're right; I'll change it and squash the changes in that same commit.

**Comment by adolfopa** [comment]
@marcogalluzzi Done. Only the lang key changed (in https://github.com/liferay-lima/liferay-portal/pull/4358/commits/39e02d1eb917cd0503de4634401fa52a41227a93); also regenerated the Build lang commit.

**Review by marcogalluzzi — APPROVED** [review]

**Comment by marcogalluzzi** [comment]
ci:forward

**Comment by marcogalluzzi** [comment]
ci:stop

**Comment by marcogalluzzi** [comment]
Hi @adolfopa the referenced PR #4356 has been merged into master. Do we proceed with this PR (remove "on hold" label) or do you want to wail until the referenced PR is available in _upstream_?

**Comment by adolfopa** [comment]
Let's wait until tomorrow. I'll rebase and push force.

**Comment by adolfopa** [comment]
Rebased and push-forced; no other changes.

**Comment by adolfopa** [comment]
ci:forward

**Comment by marcogalluzzi** [comment]
ci:test:stable

**Comment by Hanlf** [comment]
ci:forward

**Comment by Hanlf** [comment]
ci:forward:force

**Comment by marcogalluzzi** [comment]
ci:forward:force

**Comment by Hanlf** [comment]
forward manually to BChan since the ci:forward:force seems doesn't wrok: https://github.com/brianchandotcom/liferay-portal/pull/136082

**Pull request closed without merging**